main/src/ca/uwaterloo/flix/language/phase/Safety.scala

```diff
@@ -433,10 +435,11 @@ object Safety {
   }
 
   /** Checks if `cast` is legal. */
-  private def checkCheckedTypeCast(cast: Expr.CheckedCast)(implicit sctx: SharedContext, flix: Flix): Unit = cast match {
+  private def checkCheckedTypeCast(cast: Expr.CheckedCast)(implicit sctx: SharedContext, root: Root, flix: Flix): Unit = cast match {
     case Expr.CheckedCast(_, exp, tpe, _, loc) =>
-      val from = exp.tpe
-      val to = tpe
+      // Attempt to unpack the associated type. Fall back to the original type if unsuccessful.
```

Comment:
You say "unpack the associated type" but neither `from` nor `to` has to be an associated type, so the comment is a bit misleading.